libvast/src/system/connector.cpp

```diff
@@ -0,0 +1,168 @@
+    case caf::sec::stream_init_failed:
+    case caf::sec::invalid_stream_state:
+    case caf::sec::unhandled_stream_error:
+    case caf::sec::bad_function_call:
+    case caf::sec::feature_disabled:
+    case caf::sec::cannot_open_file:
+    case caf::sec::socket_invalid:
+    case caf::sec::socket_disconnected:
+    case caf::sec::socket_operation_failed:
+    case caf::sec::unavailable_or_would_block:
+    case caf::sec::malformed_basp_message:
+    case caf::sec::serializing_basp_payload_failed:
+    case caf::sec::redundant_connection:
+    case caf::sec::remote_lookup_failed:
+    case caf::sec::no_tracing_context:
+    case caf::sec::all_requests_failed:
+    case caf::sec::field_invariant_check_failed:
+    case caf::sec::field_value_synchronization_failed:
+    case caf::sec::invalid_field_type:
+    case caf::sec::unsafe_type:
+    case caf::sec::save_callback_failed:
+    case caf::sec::load_callback_failed:
+    case caf::sec::conversion_failed:
+    case caf::sec::connection_closed:
+    case caf::sec::type_clash:
+    case caf::sec::unsupported_operation:
+    case caf::sec::no_such_key:
+    case caf::sec::broken_promise:
+    case caf::sec::connection_timeout:
+    case caf::sec::action_reschedule_failed:
+      return false;
+    case caf::sec::incompatible_versions:
+    case caf::sec::incompatible_application_ids:
+      return true;
+  }
+  return true;
+}
+
+bool is_non_recoverable_error(const caf::error& err) {
+  if (err.category() != caf::type_id_v<caf::sec>)
+    return false;
+  const auto err_code = std::underlying_type_t<caf::sec>{err.code()};
+  auto err_enum = caf::sec{caf::sec::none};
+  if (!caf::from_integer(err_code, err_enum)) {
+    VAST_WARN("unable to retrieve error code for a remote node connection "
+              "error:{}",
+              err);
+    return false;
+  }
+  return is_non_recoverable_error_enum(err_enum);
+}
+} // namespace
+
+connector_actor::behavior_type
+connector(connector_actor::stateful_pointer<connector_state> self,
+          std::size_t max_connection_attempts,
+          caf::timespan connection_retry_delay, vast::port port,
+          std::string host, caf::timespan connection_timeout) {
+  self->state.connection_retry_delay = connection_retry_delay;
+  self->state.port = std::move(port);
+  self->state.host = std::move(host);
+  self->state.node_connection_timeout = connection_timeout;
+  return {[self,
+           max_connection_attempts](atom::connect) -> caf::result<node_actor> {
+            if (self->state.promise.pending())
+              return self->state.promise;
+            self->state.remaining_connection_attempts = max_connection_attempts;
+            self->state.promise = self->make_response_promise<node_actor>();
+            VAST_INFO("client connects to VAST node at {}", self->state.host);
+            self->send(self, atom::internal_v, atom::start_v);
+            return self->state.promise;
+          },
+          [self](atom::internal, atom::start) {
```

Comment:
Do you really need this second handler? It looks like you can do everything within a single handler, with the following logic:

1. Try connecting immediately
2. On success, return the node actor
3. On failure, create a response promise
4. Send yourself the same message again.

Reply:
I have tried implementing it.
I found this version less readable, but maybe it's because i didnt implement it properly.

What i don't like is:
More complicated logic on entry to atom::connect because it is now called on each retry. And the actor state must be initialized only if no "run" is in progress.

I have to use caf::delayed_anon_send to ignore the returned response promise in atom::connect. This seems like a workaround.

I think the logic is simpler with two handlers. I don't think splitting logic into multiple handlers in this case is a bad thing.
We have one handler that setups the run, and the other as implementation detail for retry.

Similar to public and private methods with difference being that the "implementation" handler can be also called by mistake :D

I have pushed a commit to show my implementation of single handler

Reply:
Maybe I've underestimated the complexity, but my naive thinking [in the comment down below](https://github.com/tenzir/vast/pull/2835#discussion_r1061841615) looks rather simple.